sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -461,6 +421,66 @@ object CollationTypeCoercion {
         else right
     }
   }
+
+  /**
+   * Throws an analysis exception if the new data type has indeterminate collation,
+   * and the expression is not allowed to have inputs with indeterminate collations.
+   */
+  private def checkIndeterminateCollation(expression: Expression, newDataType: DataType): Unit = {
+    if (shouldFailWithIndeterminateCollation(expression, newDataType)) {
+      expression.failAnalysis(
+        errorClass = "INDETERMINATE_COLLATION_IN_EXPRESSION",
+        messageParameters = Map("expr" -> toSQLExpr(expression)))
+    }
+  }
+
+  /**
+   * Returns whether the given expression which isn't allowed to have inputs with indeterminate
+   * collations has indeterminate collation.
+   */
+  private def shouldFailWithIndeterminateCollation(expression: Expression): Boolean = {
+    def getDataTypeSafe(e: Expression): DataType = try {
+      e.dataType
+    } catch {
+      case _: Throwable => NullType
+    }
+
+    expression.children.exists(child => expression.resolved &&
```

Comment:
Good catch, this was supposed to check if the child is resolved! 

However we don't need this at all since we are only running the coercion if the children are resolved:
https://github.com/apache/spark/blob/18bbddd2fe2dba04ab311e35ebdc6d69614c170f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala#L25